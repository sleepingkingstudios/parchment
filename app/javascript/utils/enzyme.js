import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
