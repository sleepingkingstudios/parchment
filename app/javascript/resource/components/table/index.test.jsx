import React from 'react';
import { shallow } from 'enzyme';

import IndexPageTable from './index';

describe('<IndexPageTable />', () => {
  const columns = jest.fn(() => ([{ label: 'Name', prop: 'name' }]));
  const data = { widgets: [] };
  const resourceName = 'widgets';
  const defaultProps = {
    columns,
    data,
    resourceName,
  };

  describe('with default props', () => {
    const rendered = shallow(<IndexPageTable {...defaultProps} />);

    it('should render a DynamicTable', () => {
      expect(rendered).toHaveDisplayName('DynamicTable');
    });

    it('should map the data', () => {
      const expected = data[resourceName];

      expect(rendered).toHaveProp({ data: expected });
    });

    it('should set the empty message', () => {
      const expected = `There are no ${resourceName} matching the criteria.`;

      expect(rendered).toHaveProp({ message: expected });
    });

    it('should set the resource name', () => {
      expect(rendered).toHaveProp({ resourceName });
    });

    it('should set the table columns', () => {
      const expected = columns();

      expect(rendered).toHaveProp({ columns: expected });
    });

    it('should match the snapshot', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with emptyMessage: string', () => {
    const emptyMessage = "These aren't the widgets you're looking for.";
    const rendered = shallow(<IndexPageTable {...defaultProps} emptyMessage={emptyMessage} />);

    it('should set the empty message', () => {
      expect(rendered).toHaveProp({ message: emptyMessage });
    });
  });

  describe('with mapData: function', () => {
    const wrappedData = { scoped: data };
    const mapData = raw => raw.scoped[resourceName];
    const rendered = shallow(
      <IndexPageTable {...defaultProps} data={wrappedData} mapData={mapData} />,
    );

    it('should map the data', () => {
      const expected = data[resourceName];

      expect(rendered).toHaveProp({ data: expected });
    });
  });

  describe('with useDestroyRequest: function', () => {
    const useDestroyRequest = jest.fn();

    beforeEach(() => { columns.mockClear(); });

    it('should pass the request to columns', () => {
      shallow(<IndexPageTable {...defaultProps} useDestroyRequest={useDestroyRequest} />);

      expect(columns).toHaveBeenCalledWith({ useDestroyRequest });
    });
  });
});
