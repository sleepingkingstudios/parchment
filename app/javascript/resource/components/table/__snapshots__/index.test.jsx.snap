// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexPageTable /> with default props should match the snapshot 1`] = `
<DynamicTable
  cellProps={Object {}}
  className={null}
  columns={
    Array [
      Object {
        "label": "Name",
        "prop": "name",
      },
    ]
  }
  data={Array []}
  message="There are no widgets matching the criteria."
  resourceName="widgets"
/>
`;
